src/components/layout/Layout.ts

```diff
@@ -136,6 +136,41 @@ export class Layout {
     }
     this.elements = [];
   }
+
+  // Duplicate element in the selected cell
+  // It will either be placed in the hoveredCell, if it is not a wall and not already occupied
+  // Or it will be placed in the first empty cell
+  duplicateElement(
+    selectedCell: [number, number],
+    hoveredCell: [number, number] | undefined,
+  ) {
+    const selectedCellY = selectedCell[0];
+    const selectedCellX = selectedCell[1];
+
+    if (!(this.layout[selectedCellY][selectedCellX] instanceof SquareType)) {
+      throw new Error('Cannot rotate a wall element');
+    }
+
+    const square = this.layout[selectedCellY][selectedCellX] as SquareType;
+
+    if (hoveredCell && this.isEmptySquare(hoveredCell[0], hoveredCell[1])) {
+      this.setElement(hoveredCell[0], hoveredCell[1], square.clone());
+    } else {
+      for (let i = 0; i < this.height * 2 - 1; i++) {
+        for (let j = 0; j < this.width * 2 - 1; j++) {
+          if (this.isEmptySquare(i, j)) {
+            this.setElement(i, j, square.clone());
+            return;
+          }
+        }
+      }
+    }
+  }
+
+  isEmptySquare(y: number, x: number): boolean {
```

Comment:
The `y` / `x` axes are correct, but I think it's better to use `i` / `j` since that's what everything else uses. It's a bit confusing to see `y` / `x` suddenly. (Same with `selectedCellY` and `selectedCellX` above)

Reply:
I disagree as i and j don't have meaning but since I'm just helping out I'll keep it consistent to the codebase.

Reply:
Fair enough, I do think there are a couple of things that need renaming so we might do that in a separate PR